Applications/util/ucom.1
.TH ucom 1 "FUZIX System Utilities" "FUZIX"
.SH NAME
ucom \- simple serial console.
.SH SYNOPSIS
.B ucom
[\-s speed] device
.SH DESCRIPTION
Simple and minimal terminal emulator program, to use with modems and alike.
.P
Just like other terminal emulator progrms
.B ucom
supports escape sequence to execute a command in
.B ucom
itself.
.SH OPTIONS
.TP
.B "-s speed"
Set speed of the
.B device
in baud before connecting.

.SH COMMANDS
Commands are executed by first inputing
.I escape character
(
.B ESC
) which by default is
.B ^A.

.TP
.B "ESC q"
Quit program and restore
.B device termios
to the previous settings.

.TP
.B "ESC ESC"
Print escape character verbatim.

.SH EXAMPLES
.P
ucom /dev/tty2
.P
ucom -s 300 /dev/tty2
.SH BUGS
To be found.
.SH AUTHORS
Yaroslav Veremenko
.SH SEE ALSO


RELEASE-NOTES

			Fuzix 0.3 Release Notes

Fuzix 0.3 is the third major release of the Fuzix OS. This release has mostly
been focussed upon kernel improvements and supporting more platforms,
particularly ZX Spectrum derived ones. The 68000 code base is now usable
but not entirely debugged and optimized. 8080 and 8085 support has been added.

Documentation

There is now a documentation directory and rules to build it
into something. It's still very incomplete and badly formatted

Warning

Fuzix 0.4 will break ABI compatibility big time. There are a pile of bad
and historic design decisions and behaviours that need to be dealt with.


User Space Changes:

New Commands: tty, vile, sok, fweeplet (Zork engine), fsh (sh with editing)
Main Improved commands:
	cp,mv,ln now taken from Heirloom so full Unix style commands
	dd can use stdin/stdout
	fweep now generally available
	sh now correctly handles out of memory
	ucp now behaves correctly if you ls a file.
	sort is now based on a different smaller codebase.
	fsck check on a clean file system is much faster.


Behavioural Changes:
- Some systems had fd first not hd first. This has been unified so on
  platforms where 0 was fd0, it's now 256 and hda is 0. If you use the fd
  or hd names at boot prompts then it's not visible

Systems Added

Amstrad PCW 8256 with CF adapter (minimal early port)
Bill Shen
	SBC64/MBC64
	Simple80 (in test)
	Tiny68K (for 68K development - very glitchy still)
Cromemco with 16FDC and 8" disks
Easy-Z80
Grant Searle Z80 CP/M design with small modifications
Linc80 with extra 16K RAM card
Pentagon 1024 with NemoIDE
RC2014 with 6502 Processor (emulation only board debug in process)
RC2014 with 8085 Processor
Sam Coupe (minimal port only at this point), Atomlite IDE
Small Computer Central
	SC108	Z80 128K RAM, CF
	SC111	Z180, 512K RAM, CF
	SC114	Z80 128K RAM, CF
SBC2G (Grant Searle style system with banked memory)
Scorpion with NemoIDE
Scrumpel (Z180)
SmallZ80
Timex TC2068/TS2068 with DivIDE and Fuzix on a cartridge
Tom's SBC (with small mods)
Video Genie with EG64B banker (and in theory TRS80 + Lubomir soft banker)
Z80-MBC2
Z80 Membership card
ZX Spectrum 128K with DivIDE or DivMMC
ZX Spectrum +3 with ZXMMC or similar

Existing Supported Systems

Amstrad NC100 with 1MB SRAM card
Amstrad NC200 with 1MB SRAM card and floppy disk drive
Dragon 32/64 with Spinx or MOOH cartridge
DX Designs P112 with 1MB RAM and G-IDE
EACA Video Genie (with suitable banked memory)
LNW Research LNW80 with suitable banked memory (TRS80 Model 1 compatible bits)
Memotech MTX512 with SDX disk (no Rememorizer or Rememotech support)
Microbee 256TC or Premium with 128K+ RAM and ideally hard disk
Multicomp09 with at least 128K RAM
RBC(*) Mark IV with optional PropIO V2
RBC(*) SBC v2 with optional PPIDE and/or PropIO V2
RC2014 SBC with 512K ROM/RAM, CF, SIO and RTC
RC2014 SBC with banked ROM, 64K RAM, CF, SIO and RTC
SOCZ80 (128MHz FPGA Z80 platform)
Tandy COCO2 with Cloud 9 IDE, or COCOSDC card and Fuzix partly in cartridge
Tandy COCO3 with suitable disk interface
TRS80 Model I/III with a hard disk and a supported banked RAM expansion
TRS80 Model 4/4D/4P 128K RAM
Zeta
Zeta V2

(*) Formerly N8VEM now RBC (RetroBrewComputers)


Incomplete Ports

Apple IIc	-	initial investigations
Atari ST	-	in progress
EZRetro		-	minimal development port for ez80
Gemini		-	early design
Lucas Nascom	-	early design
MSX1		-	works in very limited configuration only
Pentagon	-	needs the 0.4 ABI changes to be possible
ZX Uno		-	for now run the 128K ZX Spectrum version as this also
			knows about some Uno features
Z80BIOS		-	experiment to see if things like S100 can be supported
			by having a Z80 BIOS akin to the CP/M BIOS etc.
Z280RC/ZZ80RC	-	early sketches


New Virtual Platforms (Emulation Platforms For Fun or Development)

v8080		-	8080 development environment
v85		-	8085 development environment

Existing Virtual Platforms

V65C816		-	65C816 emulation work
V68		-	68000 development work
Z80Pack		-	Very flexible Z80 banked system emulator

Obsolete

MSX2		-	needs a major rework and clean up

Major Changes To Existing Platforms

Dragon-MOOH	-	now it's own port with full MOOH paging support.
MTX		-	support for CFX-I CFX-II, some rememorize support,
			base MTXPlus support, extra keys on PC adapter. Lots
			of bug fixes.

The Library/Application support has been built for
6502			no FP
65C816			no FP, using 6502 modes
6809			FP needs debug
68000			FP needs debug
68HC11			test only
8086			test only
MSP430X
NS32k			test only
PDP11			test only
Rabbit2000		test only
Z80
Z180
eZ80			needs a very recent SDCC


Core Changes

- New memory banking model for 16K banks with a fixed common
- Timer handling optimized massively.
- Correct handling of platforms with an optional RTC when it is not present
  (previously clock would drift or not run)
- Correct problems if the console had carrier detect and could be hung up
- Fix races in carrier and hangup logic
- Clear error bits correctly on Z80 SIO
- Printer ioctls added
- 6502 platform support now complete including signal handling
- eZ80 support
- Z80 memory model support for machines with no usable common RAM
- Z80 switch to a software interrupt disable tracker rather than doing the
expensive NMOS Z80 workaround
- Z80 memory model support for 32K/32K split
- 8080/85 platform support complete
- 68000 flat memory model
- /dev/sys ioctl framework for platform / cpu specific ioctls
- Unified lots of drivewire code
- Support for "swapon"
- Single process in memory systems not run parent first after fork. That
  usually means that we hit waitpid() and avoids an extra set of swaps.
- Base support for Z180 CPU dropped into a Z80 environment
- Clock set from root fs superblock timestamp on start up. This avoids clocks
  going backwards and usually means the date is right when you boot and get
  asked.
- Use uint_fast8_t for platforms where uint8_t is slow, or the compiler sucks
  at handling them.
- Move repeated swap helper into partition code.

Applications/util/ucom.c
/* ucom.c - Simple serial console.
 *
 * Copyright (C) 2024 Yaroslav Veremenko, All rights reserved.
 *
 * This file is part of FUZIX Operating System.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <stdio.h>
#include <signal.h>
#include <termios.h>
#include <unistd.h>
#include <stdlib.h>
#include <fcntl.h>
#include <string.h>
#include <errno.h>
#include <sys/wait.h>
#include <getopt.h>

static struct termios termsave[2];
static struct termios termcur[2];
static int remotefd = -1;
static pid_t child;

static int baud[] = {
    50,     /* B50 */
    75,     /* B75 */
    110,    /* B110 */
    134,    /* B134 */
    150,    /* B150 */
    300,    /* B300 */
    600,    /* B600 */
    1200,   /* B1200 */
    2400,   /* B2400 */
    4800,   /* B4800 */
    9600,   /* B9600 */
    19200,  /* B19200 */
    38400,  /* B38400 */
    57600,  /* B57600 */
    115200, /* B115200 */
};

static speed_t speed[] = {
    B50,
    B75,
    B110,
    B134,
    B150,
    B300,
    B600,
    B1200,
    B2400,
    B4800,
    B9600,
    B19200,
    B38400,
    B57600,
    B115200,
};

#ifndef uint_fast8_t
#define uint_fast8_t char
#endif

#define TTY 0
#define REMOTE 1

static void term_raw(int fd, int type)
{
    tcgetattr(fd, &termsave[type]);
    memcpy(&termcur[type], &termsave[type], sizeof(struct termios));
    cfmakeraw(&termcur[type]);
    termcur[type].c_cc[VMIN] = 1;
    termcur[type].c_cc[VTIME] = 0;
    tcsetattr(fd, TCSAFLUSH, &termcur[type]);
}

static void restore()
{
    tcsetattr(0, TCSAFLUSH, &termsave[TTY]);
    if (remotefd >= 0)
    {
        tcsetattr(remotefd, TCSAFLUSH, &termsave[REMOTE]);
        close(remotefd);
    }
}

static void quit()
{
    int status;
    kill(child, SIGHUP);
    wait(&status);
}

static int parsespeed(char* str, speed_t* s){
    register int i;
    register int b = atoi(optarg);
    for (i = 0; i < sizeof(baud) / sizeof(baud[0]); i++)
    {
        if (baud[i] == b)
        {
            *s = speed[i];
            return 1;
        }
    }
    return 0;
}

static void usage()
{
    fputs("Micro terminal program.\n"
        "Usage: ucom [-s SPEED] DEVICE\n\n"
        "  DEVICE\t- tty or other character device\n"
        "  -s SPEED\t- set device speed in baud\n"
        "  ^A\t\t- default escape character\n"
        "  ESC q\t\t- exit terminal\n", stderr);
}

int main(int argc, char *argv[])
{
    speed_t speedval = 0;
    char escchar = 1;
    int opt;
    char *remotename;

    if (isatty(0) != 1)
    {
        fputs("fd 0 is not tty\n", stderr);
        exit(1);
    }

    while ((opt = getopt(argc, argv, "s:")) != -1)
    {
        switch (opt)
        {
        case 's':
            if(!parsespeed(optarg, &speedval))
            {
                fprintf(stderr, "invalid speed: %s\n", optarg);
                exit(1);
            }
            break;
        default:
            usage();
            exit(1);
            break;
        }
    }

    argv += optind;
	argc -= optind;

    if (argc != 1)
    {
        usage();
        exit(1);
    }
        
    remotename = *argv;
    remotefd = open(remotename, O_RDWR | O_NOCTTY);

    if (remotefd < 0)
    {
        perror(remotename);
        exit(1);
    }

    term_raw(0, TTY);
    term_raw(remotefd, REMOTE);

    if (speedval > 0)
    {
        if (cfsetospeed(&termcur[REMOTE], (speed_t)speedval) < 0 ||
            tcsetattr(remotefd, TCSAFLUSH, &termcur[REMOTE]) < 0)
        {
            restore();
            perror("speed");
            exit(1);
        }
    }

    child = fork();
    if (child < 0)
    {
        restore();
        perror("fork");
        exit(1);
    }
    if (child == 0)
    {
        uint_fast8_t r;
        while (read(remotefd, &r, 1) == 1)
        {
            write(0, &r, 1);
        }
        fputs("Disconnected\r\n", stderr);
    }
    else
    {
        uint_fast8_t is_esc = 0;
        uint_fast8_t w;
        atexit(restore);
        fprintf(stderr, "Connected. Escape character ^%c\r\n", '@' + escchar);
        while (read(0, &w, 1) == 1)
        {
            if (!is_esc && w == escchar)
            {
                is_esc = 1;
                continue;
            }
            if (is_esc && w == 'q')
            {
                break;
            }
            is_esc = 0;
            write(remotefd, &w, 1);
        }
        quit();
    }
}